stdlib/public/core/StringGutsRangeReplaceable.swift

```diff
@@ -94,20 +94,28 @@ extension _StringGuts {
     if let currentCap = self.uniqueNativeCapacity, currentCap >= n { return }
 
     // Grow
-    self.grow(n) // TODO: no factor should be applied
+    self.grow(n)
   }
 
   // Grow to accommodate at least `n` code units
   @usableFromInline
   internal mutating func grow(_ n: Int) {
-    defer { self._invariantCheck() }
+    defer {
+      self._invariantCheck()
+      _internalInvariant(
+        self.uniqueNativeCapacity != nil && self.uniqueNativeCapacity! >= n)
+    }
 
     _internalInvariant(
       self.uniqueNativeCapacity == nil || self.uniqueNativeCapacity! < n)
 
-    // TODO: Don't do this! Growth should only happen for append...
-    let growthTarget = Swift.max(n, (self.uniqueNativeCapacity ?? 0) * 2)
+    // We're non-unique OR smaller than requested, so as we copy to new storage,
+    // choose a size that at least stores the current contents.
+    let growthTarget = Swift.max(n, self.utf8Count)
```

Comment:
This is not right -- the  `* 2` currently serves a crucial role, and removing it without reviewing and updating all callers would result in an algorithmic slowdown.

`grow` is currently being called from both `reserveCapacity`/`init(_initialCapacity:)` _and_ `replaceSubrange`/`append`. In the `reserveCapacity` case, we want to allocate precisely as much memory as desired (still accepting if malloc gives us more, of course), while in the `replaceSubrange` case we **must** grow storage exponentially (if the requested capacity is still within the same order of magnitude than the current size).

For example, the loop below is heavily relying on this logic so that it will only reallocate storage O(log(_n_)) times:

```swift
var s = ""
for _ in 0 ..< n {
  s.insert("a", at: s.endIndex)
}
```

Getting rid of the doubling would make this loop resize string storage O(_n_) times, rendering its runtime performance accidentally quadratic instead of linear.

(I'm using `insert` in the example because `append` is a special case: some of its code paths have their own exponential resizing code, _in addition to_ calling `grow`. `String`'s storage sizing logic has become way too obfuscated over the years; it could use a full audit and cleanup.)